mlflow/metrics/utils/make_genai_metric.py

```diff
@@ -0,0 +1,218 @@
+                "on one critical aspect. "
+                "- Score 4: the answer correctly answer the question and not missing any "
+                "major aspect"
+            ),
+            examples=[example],
+            version="v1",
+            model="gateway:/gpt4",
+            variables=["ground_truth"],
+            parameters={"temperature": 1.0},
+            aggregations=["mean", "variance", "p90"],
+            greater_is_better=True,
+        )
+    """
+
+    def eval_fn(eval_df: Union["pd.DataFrame", "pyspark.sql.DataFrame"]) -> MetricValue:
+        """
+        This is the function that is called when the metric is evaluated.
+        """
+
+        class_name = f"mlflow.metrics.utils.prompts.{version}.EvaluationModel"
+        try:
+            evaluation_model_class_module = _get_class_from_string(class_name)
+        except ModuleNotFoundError:
+            raise MlflowException(
+                f"Failed to find evaluation model for version {version}."
+                f"Please check the correctness of the version",
+                error_code=INVALID_PARAMETER_VALUE,
+            ) from None
+        except Exception as e:
+            raise MlflowException(
+                f"Failed to construct evaluation model {version}. Error: {e!r}",
+                error_code=INTERNAL_ERROR,
+            ) from None
+
+        evaluation_context = evaluation_model_class_module(
+            name,
+            definition,
+            grading_prompt,
+            examples,
+            model,
+            parameters,
+        ).to_dict()
+
+        outputs = eval_df["prediction"].tolist()
+        inputs = eval_df["input"].tolist()
+        eval_model = evaluation_context["model"]
+        eval_parameters = evaluation_context["parameters"]
+
+        # TODO: Save the metric definition in a yaml file for model monitoring
+
+        messages = []
+        for indx, (input, output) in enumerate(zip(inputs, outputs)):
+            variable_string = _format_variable_string(variables, eval_df, indx)
+            messages.append(
+                [
+                    evaluation_context["eval_prompt"].format(
+                        input=input, output=output, variables=variable_string
+                    ),
+                ],
+            )
+
+        # TODO: load the model for openai and gateway URI
+        if isinstance(eval_model, str):
+            eval_model = _load_model_or_server(eval_model, _EnvManager.LOCAL)
+        elif isinstance(eval_model, PyFuncModel):
+            pass
+        else:
+            raise MlflowException(
+                message="The model argument must be a string URI referring to an MLflow model or "
+                "an instance of `mlflow.pyfunc.PyFuncModel`.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+        # TODO: Add batch processing for messages here
+        eval_result = eval_model.predict(messages, eval_parameters)
```

Comment:
this part will be replaced by https://github.com/mlflow/mlflow/pull/9538 right? this current solution looks okay and lets us test this PR.

Reply:
Yes. The plan is to merge this for now and then update it to use open-ai and gateway instead. Will create a new PR once both of them are merged :)